### DISCUSSION TITLE
Laravel 11.23.5 (suddenly 1423 phpstan errors)

### DISCUSSION THREAD

**Discussion by EliW** [larastan/larastan#2041] [Q&A] [answered]
So wasn't sure the exact best place to post this, because I know there's been phpstan updates, and larastan updates, and laravel updates, and perhaps this should be discussed over in @calebdw's fork instead ... but, starting here

We successfully had our app running with zero phpstan errors, level: 5, against Laravel 11.20.0 ... I had switched to Caleb's fork per the other couple big threads (https://github.com/larastan/larastan/pull/1990) happening because otherwise we would have been stuck at Laravel 11.14.0 on 'stock' phpstan ...

I just sat down to test the Switch PR we had to move us from 11.20.0 to 11.23.5 ... and suddenly level:5 is giving 1423 errors to me.    These errors are centered around two general themes:

1. (The Majority) ... Models having all their properties undefined, IE:  
```
  499    Access to an undefined property App\Models\Tenant::$full_name.                        
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property  
```
Seems that basically, suddenly, phpstan/larastan aren't actually seeing any of the properties as being defined on any of our models?

2. (also) ... Relations not existing?   I have a number of 'odd' errors like this:
```
  84     Call to an undefined method Illuminate\Database\Eloquent\Relations\Relation<*, *, *>::with().            
  91     Call to an undefined method Illuminate\Database\Eloquent\Relations\Relation<*, *, *>::with().            
  94     Call to an undefined method Illuminate\Database\Eloquent\Relations\Relation<*, *, *>::select().          
```

Anyone else running into these issues?  And anyone else know what is going on?   I know I saw some things in the Laravel release notes about "Better phpstan compatibility" being included in recent releases.   Any idea what suddenly has happened?

**Comment by calebdw** [comment]
There's an open issues for the Relation ones: https://github.com/larastan/larastan/issues/2040

Also, make sure you're on the latest PHPStan 1.12.x

**Reply by calebdw** [reply]
The relation forwarding issue should be fixed on my fork

**Comment by EliW** [comment]
@calebdw thanks.   I'm on (and have been on) PHPstan 1.12.3

**Comment by EliW** [comment]
@calebdw Decided to go 'step by step'.    And realized I Was wrong, I was earlier on 1.11.x of phpstan.

Sticking with 11.20.0 right now (which I was already on) but updating all other packages.   Including:
`calebdw/larastan (v2.10.4 => v2.10.5):`
`phpstan/phpstan (1.11.10 => 1.12.3)`

It's a bunch of "int doesn't take bool" and some odd "can't call ->format() on a string" stuff, that I'm guessing is all coming from the phpstan update.

Looking into these, guessing they may be legit, but also these are sounding like larastan-ish things, as, for example, they are `tinyint` fields we are assigning boolean values to.   The the ->format() are DateTimes, that I'm guessing phpstan thing are string due to something in the config.    Going to dig deeper into these ... but completely different issue than what I was seeing at 11.23.5

**Reply by calebdw** [reply]
Sometimes the process of making PHPStan smarter means that more errors are reported, let me know if you have any questions about anything specific

**Reply by EliW** [reply]
Sorry for 'smaller updates' here @calebdw ... I just keep trying something, and expecting to 'get farther', and failing.

So basically, staying at 11.20.0 (or it would seem, 11.21 and 11.22) ... but updating phpstan.    I get these new errors, which are based on the `const CASTS` I have set up.

IE: I am getting:
`  179    Property App\Models\Policy::$cancelled (int) does not accept true.                 `
But in App\Models\Policy, I have
```
    const CASTS = [
        [...]
        'cancelled' => 'boolean',
```
and
```
    /**
     * @return self::CASTS
     */
    protected function casts(): array
    {
        return self::CASTS;
    }
```
which was all done to make earlier versions understand the types of things, per some feedback from you.

So it seems the switch up to phpstan 1.12.3 --- But without moving to Laravel 11.23, causes this 'bug'.       

Moving to Laravel 11.23, suppresses these errors (correctly) ... but gives all those other issues I was seeing.

**Reply by calebdw** [reply, marked as answer]
Try using the `$casts` property instead of the method and see if that helps:
```php
protected $casts = [
        // ...
        'cancelled' => 'boolean',
];
```

**Reply by EliW** [reply]
Will try that (but I'll swear that just months ago, we switched to this method because it made phpstan happy and the `protected costs`) didn't.   :-/

**Reply by EliW** [reply]
OK @calebdw so that fixed the new 1.12.3 errors ... though frustrating that months ago I'd just fixed this where (IIRC) ... I'm pretty sure that what I remember now was that "back then" the code had just been using the function.    And the solution to still use the 'new method' way of doing this instead of relying on the old-property ... was to move the declarations into that `const` so that I could typehint it to phpstan, and that worked.   Why did it stop working?    In any case, that switch worked.

I then updated to 11.21, no new errors.

I then updated to 11.22, no new errors.

I crossed fingers and went back to 11.23 ... 
... and now it works as well.

So I'm guessing that basically 99% of the issues were just that somehow between larastan/phpstan updating to 1.12.3 --- that it stopped looking at the cast return value of the `protected casts()` function, when it was looking at it before?   And so now it thought all the properties completely vanished.

Offhand any reason for that?  I have 'some' concern that just to get larastan to work, I'm having to do "non-suggested-laravel-coding-techiques" by using that $casts property instead of the method that the 11.x documentation says to use.

**Reply by calebdw** [reply]
> I'm having to do "non-suggested-laravel-coding-techiques" by using that $casts property instead of the method that the 11.x documentation says to use.


The `$casts` property has been used in Laravel for years and even when you declare the `casts()` method it's just being merged with the property at instantiation

**Reply by EliW** [reply]
@calebdw Yeah but the latest docs don't mention it at all and says to use the casts() method.   And I'm pretty sure (don't quote me) that Laravel Switch even therefore converts from the property to the method when converting codebases from 10.x to 11.x

Anyway, it works.   Just means that it's code that doesn't match the "docs", and isn't mentioned at all in the docs now.   So likely every new dev that walks into the project will go "wtf, this is wrong" and try to fix it, breaking phpstan reports in the process ;)
